lib/gringotts/gateways/monei.ex

```diff
@@ -435,17 +501,72 @@ defmodule Gringotts.Gateways.Monei do
       :error,
       Response.error(
         code: error.id,
-        reason: :network_fail?,
+        reason: "network related failure",
         description: "HTTPoison says '#{error.reason}'"
       )
     }
   end
 
+  defp expand_params(params, action_type) do
+    Enum.reduce_while(params, [], fn {k, v}, acc ->
+      case k do
+        :currency ->
+          if valid_currency?(v),
+            do: {:cont, [{:currency, v} | acc]},
+            else: {:halt, {:error, "Invalid currency"}}
+
+        :customer ->
+          {:cont, acc ++ make("customer", v)}
+
+        :merchant ->
+          {:cont, acc ++ make("merchant", v)}
+
+        :billing ->
+          {:cont, acc ++ make("billing", v)}
+
+        :shipping ->
+          {:cont, acc ++ make("shipping", v)}
+
+        :invoice_id ->
+          {:cont, [{"merchantInvoiceId", v} | acc]}
+
+        :transaction_id ->
+          {:cont, [{"merchantTransactionId", v} | acc]}
+
+        :category ->
+          {:cont, [{"transactionCategory", v} | acc]}
+
+        :shipping_customer ->
+          {:cont, acc ++ make("shipping.customer", v)}
+
+        :custom ->
+          {:cont, acc ++ make_custom(v)}
+
+        :register ->
+          {
+            :cont,
+            if action_type in ["PA", "DB"] do
+              [{"createRegistration", true} | acc]
+            else
+              acc
+            end
+          }
+          
+        _ ->
+          {:cont, acc}
+      end
+    end)
+  end
+
+  defp valid_currency?(currency) do
+    currency in @supported_currencies
+  end
+
   defp verification_result(%{"result" => result} = data) do
     {address, zip_code} = @avs_code_translator[result["avsResponse"]]
     code = result["code"]
-
-    results = [
+    token = data["registrationId"]
```

Comment:
The parsing job in this function can be delegated to a `separate private function` you can return a `tuple` containing the result and only do verification of the result here.
{token, common,code} = som_priv_function()